sql/src/main/java/org/opensearch/sql/sql/SQLService.java

```diff
@@ -75,6 +91,83 @@ private AbstractPlan plan(
+    if (request.format().getFormatName().equals("json")) {
+      class VisitorContext {
+        @Getter
+        @Setter
+        boolean isJSONSupported = true;
+      }
+
+      class JsonSupportedVisitor extends AbstractNodeVisitor<Boolean, VisitorContext> {
+        @Override
+        public Boolean visit(Node node, VisitorContext context) {
+          // A node is supported if all of its children are supported.
+          return node.getChild().stream().filter(c -> c.accept(this, context) != null)
+              .allMatch(c -> c.accept(this, context));
+        }
+
+        @Override
+        public Boolean visitProject(Project node, VisitorContext context) {
+          // Overridden visit functions are done in memory and are not supported with json format
+          class UnsupportedExpressionsVisitor
+              extends AbstractNodeVisitor<Boolean, Object> {
+
+            @Override
+            public Boolean visitFunction(Function node, Object context) {
+              // queries with function calls are not supported.
+              return false;
+            }
+
+            @Override
+            public Boolean visitLiteral(Literal node, Object context) {
+              // queries with literal values are not supported
+              return false;
+            }
+
+            @Override
+            public Boolean visitCast(Cast node, Object context) {
+              // Queries with cast are not supported
+              return false;
+            }
+
+            @Override
+            public Boolean visitAlias(Alias node, Object context) {
+              // Alias node is accepted if it does not have a user-defined alias
+              // and if the delegated expression is accepted.
+              if (!StringUtils.isEmpty(node.getAlias()))// && node.getDelegated().accept(this, context))
+                return false;
+              else {
+                return node.getDelegated().accept(this, context);
+              }
+            }
+          }
+
+          // Project node is supported if all of its children are supported and
+          // all expressions in project list are supported.
+          UnsupportedExpressionsVisitor unsupportedExpressionsVisitor = new UnsupportedExpressionsVisitor();
+          return visit(node, context)
+              && node.getProjectList().stream().filter(c -> c.accept(unsupportedExpressionsVisitor, context) != null)
+              .allMatch(e -> e.accept(unsupportedExpressionsVisitor, context));
+
+        }
+
+        @Override
+        public Boolean visitAggregation(Aggregation node, VisitorContext context) {
+          return node.getGroupExprList().isEmpty();
+        }
+      }
+
+      Boolean isJsonSupported = ((Query) statement).getPlan().accept(new JsonSupportedVisitor(), new VisitorContext());
+
+      if (!isJsonSupported) {
+        throw new UnsupportedOperationException(
+            "Queries with aggregation and in memory operations (cast, literals, alias, math, etc.)"
```

Comment:
Curious what of these are supported in legacy engine with JSON format?

Reply:
All of those are supported in legacy because they are pushed down.

Reply:
For slightly better error messaging, throw the errors in `JsonSupportAnalyzer` and you can say what type of functions are unsupported as the node visitor reaches an unsupported node.

Reply:
I've made each unsupported node visitor throw the exception in https://github.com/Bit-Quill/opensearch-project-sql/commit/74f88ae574b06eb63778a511830e0c148cda8539. Thanks!